COOPR:629 Update template APIs to support extending templates and including partials

rpedanov opened the pull request:
@jwang47 Branch ready for review.

Commit:
COOPR-699
base implementation

Commit:
coopr-629
rename builder

Commit:
COOPR-629
refactored merger, service-layer config

Commit:
COOPR-629
Tests for partial templates / merging fix

Commit:
COOPR-578
excluded constants

Commit:
COOPR-629
split exceptions / javadoc

Commit:
COOPR-629
not merged file fix

Commit:
COOPR-629
extend test

Commit:
COOPR-629
strange build error fix-hack

Commit:
COOPR-629
merge from develop

rpedanov commented:
@jwang47 I'm not sure why build fails, it looks like some npm problems. Could you check it please?

Commit:
COOPR-629
generic fix

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/Administration.java:
```diff
@@ -20,10 +20,10 @@
   public static final Administration EMPTY_ADMINISTRATION = new Administration(LeaseDuration.FOREVER_LEASE_DURATION);
 
-  private final LeaseDuration leaseDuration;
+  LeaseDuration leaseDuration;
```
Why remove `private final`?

rpedanov replied:
For merging purposes. We have to replace leaseDuration at the final merging process.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java:
```diff
@@ -16,32 +16,28 @@
-    //TODO: move it to the final phase
-//    Preconditions.checkArgument(provider != null, "default provider must be specified");
-    Preconditions.checkArgument(dnsSuffix == null || StringUtils.isValidDNSSuffix(dnsSuffix),
-                                dnsSuffix + " is an invalid DNS suffix.");
```
Why remove these checks?

rpedanov replied:
Since we have partial templates this fields can be empty. So I moved this checks into Merger.validateFields() that it can be validated when the template will be merged.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java:
```diff
@@ -129,12 +157,12 @@ public static Builder builder() {
 
     public Builder setServices(Set<String> services) {
-      this.services = ImmutableSet.copyOf(services);
+      this.services = services == null ? Sets.<String>newHashSet() : services;
```
`services == null ? ImmutableSet.of() : ImmutableSet.copyOf()`

rpedanov replied:
Since we have to merge several templates we have to merge all inner collections of this templates. For example if we have two templates with "defaults" for merge then we have to merge "services" set from second defaults into the first one. If set will be immutable then this operation will be no successfully.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterDefaults.java:
```diff
@@ -129,12 +157,12 @@ public static Builder builder() {
 
     public Builder setServices(String... services) {
-      this.services = ImmutableSet.copyOf(services);
+      this.services = new HashSet<String>(Arrays.asList(services));
```
Why change this to be mutable set?

rpedanov replied:
See comment above

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
+        return this;
+      }
+
+      public Merger merge(Collection<AbstractTemplate> templates) throws TemplateImmutabilityException {
```
Should be `List<AbstractTemplate>` since order is important.

rpedanov replied:
LinkedHashSet always passed here. I believe that it saves insert order as well.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
+        for (AbstractTemplate template : templates) {
+          if (template instanceof ClusterTemplate) {
+            copyFullProps(template);
+          } else {
```
Should be `else if (template Instanceof PartialTemplate)`, and throw in the `else` - if we have a new child of `AbstractTemplate`, then we won't get unexpected behavior.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
+        Preconditions.checkArgument(clusterDefaults.getProvider() != null, "default provider must be specified");
+        Preconditions.checkArgument(clusterDefaults.getDnsSuffix() == null ||
+                                      StringUtils.isValidDNSSuffix(clusterDefaults.getDnsSuffix()),
+                                    clusterDefaults.getDnsSuffix() + " is an invalid DNS suffix.");
```
Ok, so the checks are moved here.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -73,10 +83,13 @@ public Parent getParent() {
+    public Merger merger() {
+      return new Merger();
+    }
+
```
I don't think `Merger` should come from a builder method. What if we want to use multiple `Merger` implementations? I think we should instead have `Merger` be injected into whatever classes want to merge templates.

jwang47 replied:
@rpedanov Please look at this comment too.

rpedanov replied:
@jwang47 Yes, I already did it. I'm waiting when I done with moving mutable collections. Please, see my question below.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/Compatibilities.java:
```diff
@@ -133,37 +135,37 @@ public static Builder builder() {
-    private Set<String> services = ImmutableSet.of();
+    private Set<String> hardwaretypes = Sets.newHashSet();
+    private Set<String> imagetypes = Sets.newHashSet();
+    private Set<String> services = Sets.newHashSet();
```
Why change these fields to be mutable sets?

rpedanov replied:
For merging purposes. See comment above.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/Constraints.java:
```diff
@@ -28,12 +28,12 @@
                      SizeConstraint sizeConstraint) {
     this.serviceConstraints = serviceConstraints == null ?
-      ImmutableMap.<String, ServiceConstraint>of() : serviceConstraints;
+      Maps.<String, ServiceConstraint>newHashMap() : serviceConstraints;
```
Why change this?

rpedanov replied:
For merging purposes. See comment above.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
+
+      @SuppressWarnings("unchecked")
+      private void mergeSet(Set dest, Set from) {
+        if (dest != null && from != null) {
```
We made a lot of things mutable sets due to how we're merging the template here, right?

jwang47 replied:
Since we made a lot of things mutable sets/maps, let's move the immutable copy to the `build()` method of the builder, so that a built object doesn't use mutable sets/maps.

jwang47 replied:
@rpedanov Please look at this comment.

rpedanov replied:
Hey @jwang47 , as I understand I have to remove all Mutable collections and its setters from builders? But create setters for mutable collections inside template entities (ClusterDefaults, Compatibilities)?  So if we try to create some entity we have to construct it from builder first and after that init not mutable fields through setter of this entity?

jwang47 replied:
Right, let's keep the collections of the template entities (ClusterDefaults, Compatibilities) immutable as they were before. Just don't want to allow the possibility of any unexpected bugs.

jwang47 commented on coopr-server/src/test/java/co/cask/coopr/spec/template/PartialTemplateTest.java:
```diff
@@ -2,23 +2,30 @@
 /**
  *
  */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
```
Why is this needed?

rpedanov replied:
For test ordering. I want that templates will be parsed and persisted at the begin so we can get it in the next tests, and deleted at the final test method.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/spec/template/ClusterTemplate.java:
```diff
@@ -97,6 +110,122 @@ public ClusterTemplate build() {
+          administration.leaseDuration = from.getAdministration().getLeaseDuration();
+        }
+
+        //merge base entity properties
```
We don't want to merge these. Instead, we'll use whatever is immediately provided. E.g. with user JSON, we would simply use what is provided in that JSON; if it's from a specified template, then it's from that template.

rpedanov replied:
Now this fields will be set with the values of last template that be provided in merge set. I mean that if we have something like that:
Child -> Parent -> Includes
Then values of this fields will be replaced with child values because we merge child at final. This allow to avoid some null value from the child. For example if parent has Label field value and child has not, then Label will be taken from parent. Or we should take this fields only from the last template(specified template) and set null values if values does not provided?

jwang47 replied:
Let's take these fields only from the last template.

jwang47 commented:
Please resolve merge conflicts

Commit:
COOPR-629
merger refactoring / rollback immutably collections / remove private scope of template inner collections

Commit:
COOPR-629
merge with develop

Commit:
COOPR-629
resolve merge conflicts

Commit:
COOPR-629
services order changes

rpedanov commented:
Immutable collections came back to template entities, merger separate service now. @jwang47 please check it.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java:
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
+   * Build cluster template from provided includes and parents.
+   *
+   * @param clusterTemplate Cluster template which can contains includes and parents.
+   * @param account Account of the user that is trying to resume a cluster job.
```
Account of the user that is trying to resolve a template.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java:
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
+   * Build cluster template from provided includes and parents.
+   *
+   * @param templateName Cluster template name.
+   * @param account Account of the user that is trying to resume a cluster job.
```
Account of the user that is trying to resolve a template.

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java:
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
+   * @param clusterTemplate Cluster template which can contains includes and parents.
+   * @param account Account of the user that is trying to resume a cluster job.
+   * @return Cluster Template with merged body from includes and parents.
+   * @throws IOException if there was some error writing to stores.
```
We don't write to stores here, right? Just read?

jwang47 commented on coopr-server/src/main/java/co/cask/coopr/cluster/ClusterService.java:
```diff
@@ -732,4 +741,85 @@ private Cluster getCluster(String clusterId, Account account) throws IOException
+   * @param account Account of the user that is trying to resume a cluster job.
+   * @return Cluster Template with merged body from includes and parents.
+   * @throws IOException if there was some error writing to stores.
+   * @throws TemplateNotFoundException if template can't be found in store.
```
The template is provided, so why do we need to throw TemplateNotFoundException?

jwang47 commented:
Let's also add test cases for user providing an entire template rather than using a stored template. Can test the following:
- Provided template without error
- Provided template that tries to override immutable partial
- Provided template that tries to override immutable parent template
- Provided template that is badly formatted (i.e. not proper JSON)
- etc

(also make sure we have the same set of test cases that use stored templates)

Otherwise, the code changes mostly looks good.

Commit:
COOPR-629
template codec null guards
add compatibilities codec into gson adapters
template entities final links for to avoid errors
merger now merge trough builders
back num machines "1" value to ClusterServiceTest (rollback from develop)
extra template tests
java docs text fix

rpedanov commented:
@jwang47 please take a look

jwang47 commented:
Just one minor comment.

Commit:
COOPR-629
new line

rpedanov commented:
@jwang47 new line added
